### PR TITLE
[BugFix] fix union rewrite bug

### LINKED ISSUE #14900: [materialized view]unknow error when select from mv
### Steps to reproduce the behavior (Required)
```
CREATE TABLE t1 (
                    k1 INT,
                    v1 INT,
                    v2 INT)
                DUPLICATE KEY(k1)
                PARTITION BY RANGE(`k1`)
                (
                PARTITION `p1` VALUES LESS THAN ('3'),
                PARTITION `p2` VALUES LESS THAN ('6'),
                PARTITION `p3` VALUES LESS THAN ('9'),
                PARTITION `p4` VALUES LESS THAN ('12'),
                PARTITION `p5` VALUES LESS THAN ('15'),
                PARTITION `p6` VALUES LESS THAN ('18')
                )
                DISTRIBUTED BY HASH(k1);
                insert into t1 values (1,1,1),(2,1,1),(3,1,1),
                                      (4,1,1),(5,1,1),(6,1,1),
                                      (7,1,1),(8,1,1),(9,1,1),
                                      (10,1,1),(11,1,1);
CREATE MATERIALIZED VIEW mv1 DISTRIBUTED BY HASH(k1) REFRESH ASYNC AS SELECT k1,v1,v2 from t1 where k1>1;
CREATE MATERIALIZED VIEW mv2 DISTRIBUTED BY HASH(k1) REFRESH ASYNC AS SELECT k1,v1,v2 from mv1 where k1>2;
CREATE MATERIALIZED VIEW mv3 DISTRIBUTED BY HASH(k1) REFRESH ASYNC AS SELECT k1,v1,v2 from mv2 where k1>3;
select * from mv1;
```
### Expected behavior (Required)
```
+------+------+------+
| k1   | v1   | v2   |
+------+------+------+
|    5 |    1 |    1 |
|    6 |    1 |    1 |
|    8 |    1 |    1 |
|    3 |    1 |    1 |
|    4 |    1 |    1 |
|    7 |    1 |    1 |
|   11 |    1 |    1 |
|    2 |    1 |    1 |
|    9 |    1 |    1 |
|   10 |    1 |    1 |
+------+------+------+
```
### Real behavior (Required)
```
2022-12-08 16:17:30,291 WARN (starrocks-mysql-nio-pool-11|358) [StmtExecutor.execute():517] execute Exception, sql select * from mv1
java.lang.IllegalStateException: null
	at com.google.common.base.Preconditions.checkState(Preconditions.java:494) ~[spark-dpp-1.0.0.jar:?]
	at com.starrocks.sql.optimizer.statistics.Statistics.getColumnStatistic(Statistics.java:56) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.statistics.ExpressionStatisticCalculator$ExpressionStatisticVisitor.visitVariableReference(ExpressionStatisticCalculator.java:63) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.statistics.ExpressionStatisticCalculator$ExpressionStatisticVisitor.visitVariableReference(ExpressionStatisticCalculator.java:42) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator.accept(ColumnRefOperator.java:126) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.statistics.ExpressionStatisticCalculator.calculate(ExpressionStatisticCalculator.java:39) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.statistics.ExpressionStatisticCalculator.calculate(ExpressionStatisticCalculator.java:35) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.statistics.PredicateStatisticsCalculator$PredicateStatisticsCalculatingVisitor.getExpressionStatistic(PredicateStatisticsCalculator.java:317) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.statistics.PredicateStatisticsCalculator$PredicateStatisticsCalculatingVisitor.visitBinaryPredicate(PredicateStatisticsCalculator.java:195) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.statistics.PredicateStatisticsCalculator$PredicateStatisticsCalculatingVisitor.visitBinaryPredicate(PredicateStatisticsCalculator.java:31) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.operator.scalar.BinaryPredicateOperator.accept(BinaryPredicateOperator.java:62) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.statistics.PredicateStatisticsCalculator.statisticsCalculate(PredicateStatisticsCalculator.java:28) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.statistics.StatisticsCalculator.estimateStatistics(StatisticsCalculator.java:1242) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.statistics.StatisticsCalculator.visitOperator(StatisticsCalculator.java:172) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.statistics.StatisticsCalculator.computeOlapScanNode(StatisticsCalculator.java:238) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.statistics.StatisticsCalculator.visitLogicalOlapScan(StatisticsCalculator.java:201) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.statistics.StatisticsCalculator.visitLogicalOlapScan(StatisticsCalculator.java:136) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.operator.logical.LogicalOlapScanOperator.accept(LogicalOlapScanOperator.java:120) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.statistics.StatisticsCalculator.estimatorStats(StatisticsCalculator.java:152) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.task.DeriveStatsTask.execute(DeriveStatsTask.java:42) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.task.SeriallyTaskScheduler.executeTasks(SeriallyTaskScheduler.java:49) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.memoOptimize(Optimizer.java:380) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.optimizeByCost(Optimizer.java:149) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.optimizer.Optimizer.optimize(Optimizer.java:91) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.StatementPlanner.createQueryPlan(StatementPlanner.java:95) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.StatementPlanner.plan(StatementPlanner.java:66) ~[starrocks-fe.jar:?]
	at com.starrocks.sql.StatementPlanner.plan(StatementPlanner.java:37) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.StmtExecutor.execute(StmtExecutor.java:368) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.ConnectProcessor.handleQuery(ConnectProcessor.java:312) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.ConnectProcessor.dispatch(ConnectProcessor.java:427) ~[starrocks-fe.jar:?]
	at com.starrocks.qe.ConnectProcessor.processOnce(ConnectProcessor.java:673) ~[starrocks-fe.jar:?]
	at com.starrocks.mysql.nio.ReadListener.lambda$handleEvent$0(ReadListener.java:55) ~[starrocks-fe.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_252]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_252]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_252]
```
### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`
```
+----------------------------+
| current_version()          |
+----------------------------+
| BRANCH-2.5-RELEASE 52fe458 |
+----------------------------+
```

### PR DESCRIPTION
**Pull request by ABingHuang**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #14900 

## Problem Summary(Required) ：

fix union rewrite bug. the compensation predicate of query should be rewritten against query. 
fix it by rewriting the view predicate to query predicates.
## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 27 / 39 (69.23%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|26|38|68.42%|[299, 300, 301, 303, 304, 305, 306, 307, 309, 310, 312, 327]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java|1|1|100.00%|[]|

**Comment by ABingHuang** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 1]
```diff
@@ -288,12 +288,54 @@ private OptExpression tryUnionRewrite(RewriteContext rewriteContext, OptExpressi
...
 
+        ScalarOperator equalPredicates = MvUtils.canonizePredicate(compensationPredicates.getEqualPredicates());
+        ScalarOperator otherPredicates = MvUtils.canonizePredicate(Utils.compoundAnd(
+                compensationPredicates.getRangePredicates(), compensationPredicates.getResidualPredicates()));
+        Map<ColumnRefOperator, ScalarOperator> queryExprMap = MvUtils.getColumnRefMap(
+                rewriteContext.getQueryExpression(), rewriteContext.getQueryRefFactory());
+        Set<ColumnRefOperator> mvRefSets = rewriteContext.getMvRefFactory().getColumnRefToColumns().keySet();
+        mvRefSets.removeAll(rewriteContext.getQueryColumnSet());
+        if (!ConstantOperator.TRUE.equals(equalPredicates)) {
+            Multimap<ScalarOperator, ColumnRefOperator> normalizedMap =
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:299]: The logical in the two if clause are so similar, why not use a fcuntion.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java [hunk 2]
```diff
@@ -288,12 +288,54 @@ private OptExpression tryUnionRewrite(RewriteContext rewriteContext, OptExpressi
                 || !ConstantOperator.TRUE.equals(compensationPredicates.getRangePredicates())
                 || !ConstantOperator.TRUE.equals(compensationPredicates.getResidualPredicates()));
 
+        ScalarOperator equalPredicates = MvUtils.canonizePredicate(compensationPredicates.getEqualPredicates());
+        ScalarOperator otherPredicates = MvUtils.canonizePredicate(Utils.compoundAnd(
+                compensationPredicates.getRangePredicates(), compensationPredicates.getResidualPredicates()));
+        Map<ColumnRefOperator, ScalarOperator> queryExprMap = MvUtils.getColumnRefMap(
+                rewriteContext.getQueryExpression(), rewriteContext.getQueryRefFactory());
+        Set<ColumnRefOperator> mvRefSets = rewriteContext.getMvRefFactory().getColumnRefToColumns().keySet();
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java:296]: why not a `ColumnRefSet` ?

**Review by murphyatwork — DISMISSED** [review]

**Review by Astralidea — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by Youngwb — DISMISSED** [review]

**Review by Astralidea — DISMISSED** [review]

**Review by Astralidea — DISMISSED** [review]

**Comment by Astralidea** [comment]
run starrocks_fe_unittest

**Review by Youngwb — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Comment by ABingHuang** [comment]
run starrocks_admit_test

**Review by Astralidea — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by ABingHuang** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 31 / 36 (86.11%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java|28|33|84.85%|[238, 239, 300, 306, 344]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/materialization/AggregatedMaterializedViewRewriter.java|3|3|100.00%|[]|

**Comment by colorfulu** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**